Add multipath

flub opened the pull request.

MikeRomaniuk commented:
Hi there! 
Can you please say how much of the work is done already (just a ballpark figure will be enough)? 
I would like to use quinn and the multipath feature feature is very important to me. 
If this would be possible (I would need to ask the stakeholders) I may contribute to the development of the feature.

flub commented:
@MikeRomaniuk Can I inquire as to what your usecase is?

MikeRomaniuk commented:
@flub, thanks for answering.
Sorry for the long time without an answer. 
Unfortunately, I can't answer for the use case. 
I would like to join you. I think I may answer whether I could help you in a few days or so.

flub commented:
@MikeRomaniuk The usecase does kind of matter, because we have a very specific usecase in mind (documented in my fosdem talk, though the recording is terrible - i should re-record that sometime).  We want to do the absolute minimum we can do for our usecase, and this keeps the number of public APIs that will be needed small as well.

The multipath spec does give you enough to make a largely interoperable protocol, but also still leaves a lot of things as essentially further research topics.  The largest one is probably packet scheduling, also things like how to interoperate with the ack-frequency spec and probably a bunch more that escape me right now.  So what you want to get out of this does matter in that regard.

MikeRomaniuk commented:
@flub, I might have understood your speech in the video wrong, so feel free to correct me.
You want to implement multipath, so after a NAT traversal, you could be able to connect a second direct path to get around any firewalls. After the connection, the first path will serve as a backing and/or will be used to increase bandwidth.
Did I get this right?

flub commented:
@MikeRomaniuk Yes, more or less.  Our "typical" case is to establish the connection via a relayed path, we'll use an IPv6 private address space IP to identify paths via the relay and handle that in our impl of Quinn's AsyncUdpSockt trait.  This path will be set as PATH_BACKUP.   Then combine draft-seeman-quic-nat-traversal with multipath to open another path using real IP addresses and mark that path as PATH_AVAIALBE.

This will give us very simple packet scheduling semantics, I think the multipath part of this it's about the simplest version of multipath you can build.  Once we have this working stable and reliable we will probably start looking at more advanced packet scheduling, there are a lot of possible things, as you say maximise bandwith by using multiple paths, or interactions with lost packets etc.  But that is much further down the line for us, we really want to have the simplest multipath first.

MikeRomaniuk commented:
@flub, thanks for clarification. 
For the project I am about to start, the main goal of using multipath extension for QUIC is to maximise bandwith. 
Right now I am considering [tquic](https://github.com/Tencent/tquic) and your fork of quinn. I am researching the state of the available crates. 
I am curious, when do you think, you will proceed to this stage?

MikeRomaniuk commented:
@flub I am very grateful for your explanations.
My goal here is to find out with which approach should I stick with. 
Right now, I have 2 options: 
* Implement async bindings for tquic (it doesn't have them)
* Help you implement multipath extension

Honestly, I would rather use quinn over tquic, since it is widely used in the community and I believe will be a better choice in the longrun. 

I will need to talk to the team to take into account their opinion and I will let you if we would decide to proceed our way with you.

MikeRomaniuk commented:
@shirok1 thanks for the recommendation!

I am very glad that you are interested in QUIC and the problem I am about to tackle. However, I cannot invite you to participate, since this is not my personal project.

flub commented:
@MikeRomaniuk 

> I will need to talk to the team to take into account their opinion and I will let you if we would decide to proceed our way with you.

Did you progress on your evaluation?  Would be good to know either way what the outcome is.

MikeRomaniuk commented:
Hi @flub!
Sorry for a long time without an answer. Team decided that they will go with another approach.

flub commented:
> Hi @flub! Sorry for a long time without an answer. Team decided that they will go with another approach.

@MikeRomaniuk No worries, all the best!

dignifiedquire committed:
refactor: improve path ack handling

flub committed:
Use a per-path loss timer

I didn't understand this fully at first and was worried that the
places where a loss timer was set did not always correspond with the
path they should be set for.  Having reviewed this I think it's doing
the correct thing for the paths.

Another benefit is that multiple loss-detection timers can now expire
and be handled in the same poll_timers() call, which is a little more
efficient.

And it also simplifies the code a bunch, so that's good.

Maybe it is even worth making this timers per-space as well?

flub committed:
Merge branch 'multipath-quinn-0.11.x' into flub/loss-timer

flub committed:
Remove obsolete comment

flub committed:
fix merge

dignifiedquire committed:
fix(quinn-proto): correctly associate ACKs to paths (#79)

* fix(quinn-proto): correctly associate ACKs to paths

* remove path_id argument

flub committed:
Fix the LossDetection timer for multipath (#70)

dignifiedquire committed:
feat(quinn-proto): validate paths_blocked and path_cids_blocked frames

flub committed:
Keep per-path PathStats

flub committed:
Keep per-path PathStats (#80)

dignifiedquire committed:
feat(quinn-proto): handle receiving of path availability frames (#77)

* feat(quinn-proto): handle reciving of path availability frames

* apply CR

divagant-martian committed:
update config api to set concurrent open paths (#82)

* update config to use max_concurrent_multipath_paths & use noegiated vs enabled lang

* fix docs, remove outdated TODO

dignifiedquire committed:
feat: initial setup for high level open path api (#81)

* feat: initial setup for high level open path api

* fix logic

* feat(quinn): introduce `Path` struct

* implement path close and some cleanup

* fix docs

---------

Co-authored-by: Diva M <divma@protonmail.com>

gongy committed:
Catch EOPNOTSUPP as failure for opportunistic flags

flub committed:
This adds a frame stat for PATH_NEW_CONNECTION_ID

As part of this a few minor fixes, in recording and logging the
PATH_NEW_CONNECTION_ID frames.

Also a few fixes for logging frames that were mistakenly not logged.

flub committed:
This adds a frame stat for PATH_NEW_CONNECTION_ID (#83)

As part of this a few minor fixes, in recording and logging the
PATH_NEW_CONNECTION_ID frames.

Also a few fixes for logging frames that were mistakenly not logged.

flub committed:
Improve the cid_issued test

flub committed:
Improve the cid_issued test (#84)

matheus23 committed:
Store a `Box<dyn AsyncUdpSocket>` instead of `Arc`ing it, make `poll_recv` take `&mut self`

abonander committed:
expose `Connection::side()` in `quinn`

matheus23 committed:
Refactor `UdpPoller` into `UdpSender` and use it in favor of `AsyncUdpSocket` in connections

flub committed:
Ensure that a non-zero len CID is used with multipath

matheus23 committed:
refactor(quinn): Configure out `async_io::UdpSocket` when unused

flub committed:
Ensure that a non-zero len CID is used with multipath (#86)

With a test!  Check me out!

flub committed:
wip

flub committed:
this is done by the caller now

flub committed:
make it work, with collect(), yolo

stormshield-damiend committed:
perf: allow setting initial round trip time

flub committed:
Remove my logging

flub committed:
clippy

stormshield-damiend committed:
perf: allow configuring ack frequency

flub committed:
Flub/acks again (#87)

flub committed:
Move last-received packet number to the per-path data

Somehow that managed to escape so far.

Also, look at me embracing unwrap!

divagant-martian committed:
add a paths to open queue and start creating some errors (#89)

* add a paths to open queue and start creating some errors

* wip

* improve error cases when queueing a path open

* queue two frames for opening a path or failing to do so

* queue open path

* fix future polling

* ugly return types are declared ugly

* fix doc link

flub committed:
Move last-received packet number to the per-path data (#88)

Somehow that managed to escape so far.

Also, look at me embracing unwrap!

flub committed:
Start sending path abandon

This changes the open_path function to no longer return the
PATH_ABANDON frame, but instead put it on the pending data.  And adds
functionality to send it in poll_transmit.

divagant-martian committed:
rename frame's write and read methods to the more used encode and decode (#91)

flub committed:
Start sending path_available/path_backup

flub committed:
Split the frame type

flub committed:
test setting it back

flub committed:
Fix can_send for path status frames

flub committed:
Improve logging a little bit

flub committed:
PATH_ACK frames are not ack-eliciting

flub committed:
PATH_ACK frames are not ack-eliciting (#93)

flub committed:
Keep track of remote state separately

flub committed:
Start sending path abandon (#90)

This changes the open_path function to no longer return the PATH_ABANDON
frame, but instead put it on the pending data. And adds functionality to
send it in poll_transmit.

flub committed:
Fix remote path status state

flub committed:
clippy

flub committed:
send and receive PATH_AVAILABLE/PATH_BACKUP (#92)

This always accepts the status set by the peer, the draft says it is
only a suggestion.

flub committed:
Add frame stats to for path status frames

flub committed:
Do PTO calculations correct

- PTO is computed for the right path
- crypto keys update uses 3 * max PTO of all paths
- draining state uses 3 * max PTO of all paths

flub committed:
Add frame stats to for path status frames (#94)

flub committed:
Do PTO calculations correct (#95)

- PTO is computed for the right path
- crypto keys update uses 3 * max PTO of all paths
- draining state uses 3 * max PTO of all paths

abonander committed:
feat: add `Connection::set_send_window()`

flub committed:
Scaffolding for per-path idle timer

flub committed:
Handle PathIdle timer expiry

mxinden committed:
chore(udp): increase patch version to v0.5.13

flub committed:
Do keep alives

flub committed:
Allow configuring the per-path idle-timeout and keep-alive

flub committed:
Add quinn api for this

flub committed:
fix some docs

flub committed:
use correct location for docs

flub committed:
more docs fixes

stormshield-fabs committed:
collect recovery metrics with qlog

stormshield-fabs committed:
add qlog option to perf

FlorianUekermann committed:
proto: Add option to pad application data UDP datagrams to MTU

djc committed:
proto: upgrade to rustls-platform-verifier 0.6

djc committed:
Clean up unnecessary advisory ignore

flub committed:
Per-path idle timeout and keepalive (#96)

- [x] Basic idle timeout scaffolding.
- [x] Handle `PathIdle` timer expiry.
- [x] Basic keep alive scaffolding.
- [x] Handle `PathKeepAlive` timer expiry.
- [x] quinn-proto API for setting path idle and keepalive.
- [x] quinn API for setting path idle and keepalive.
- [x] Allow setting path idle & keepalive defaults in `TransportConfig`.

The following might perhaps better flow over to another PR
- [ ] Abandon the path correctly.
- [ ] Close the connection if there is no remaining path.

flub committed:
Refuse to close a path if it is the last path available.

flub committed:
clippy

flub committed:
Refuse to close a path if it is the last path available. (#97)

flub committed:
Assert we get PATH_ACK stats

flub committed:
Assert we get PATH_ACK stats (#98)

stormshield-damiend committed:
perf: allow selecting congestion algorithm

stormshield-damiend committed:
perf: leave async tasks early

divagant-martian committed:
improve reporting of frames on open path errors (#99)

divagant-martian committed:
Fix bugs in the process of opening a new path (#100)

flub committed:
Tidy up frame stats for issued/retired CIDs

flub committed:
clippy

flub committed:
Fix lints for clippy 1.88

flub committed:
Tidy up frame stats for issued/retired CIDs (#101)

Ralith committed:
Factor out ServerConfig::has_preferred_address getter

Ralith committed:
Don't violate client CID limits when advertising a preferred address

gry8t committed:
Updated quinn::Connection::send_datagram() docs with regards to full queue behaviour

flub committed:
Open paths immediately

This way the state for open paths is in fewer places.  There is no
functionality change but fewer moving pieces.

flub committed:
Remove old stuff

flub committed:
Open paths immediately (#102)

This way the state for open paths is in fewer places. There is no
functionality change but fewer moving pieces.

divagant-martian committed:
Add path id to spans to help with debugging (#103)

stormshield-fabs committed:
bump MSRV to 1.74.1

djc committed:
Upgrade hex-literal to 1

djc committed:
Upgrade rcgen to 0.14

flub committed:
Fully test behaviour of MAX_PATH_ID and CID issuing

It fixes up the state of the paths kept on the connection and attempts
to document them a bit.

The MAX_PATH_ID and any possible PATH_NEW_CONNECTION_ID frames are
unfortunately not sent in the same datagram, because the endpoint
dirver issues CIDs asynchronously.

flub committed:
Use crate::Instant for wasm

flub committed:
remove unused function

flub committed:
Consider abandoned paths too

flub committed:
PR feedback, mostly comment fixes

flub committed:
Do not try to compute max_path_id on the fly

This is already set to the right value.

flub committed:
fix off-by-one error

fun times

flub committed:
Fully test behaviour of MAX_PATH_ID and CID issuing (#104)

It fixes up the state of the paths kept on the connection and attempts
to document them a bit.

The MAX_PATH_ID and any possible PATH_NEW_CONNECTION_ID frames are
unfortunately not sent in the same datagram, because the endpoint dirver
issues CIDs asynchronously.

flub committed:
Sequence no from (PATH_)RETIRE_CONNECTION_ID is a VarInt

flub committed:
Sequence no from (PATH_)RETIRE_CONNECTION_ID is a VarInt (#105)

This is way to easy to get wrong, but I also checked with main.

flub committed:
Start sending PATH_CIDS_BLOCKED

WIP

djc committed:
Upgrade to fastbloom 0.12

Thomasdezeeuw committed:
Update to socket2 v0.6

Ralith committed:
Fix flakes in key_update_reordered

When the client's initial randomized key_update_phase happens to be
small enough, a second key update would be initiated before the first
had been completed, causing transient packet loss.

Ralith committed:
Ignore forced key updates outside established state

divagant-martian committed:
add new path (#107)

mxinden committed:
chore(Cargo.toml): support both socket2 v0.5 and v0.6

Enables quinn-udp users like Firefox to gradually upgrade to socket2 v0.6.

dignifiedquire committed:
Merge 0.11.x

dignifiedquire committed:
fixups after merge

dignifiedquire committed:
happy clippy

dignifiedquire committed:
ci: fixup

flub committed:
feat: merge iroh-0.11.x (#109)

divagant-martian committed:
Fix bugs in path status (#108)

flub committed:
Fill in the nex_seq and check in on receipt

dependabot[bot] committed:
build(deps): bump rcgen from 0.14.0 to 0.14.1

Bumps [rcgen](https://github.com/rustls/rcgen) from 0.14.0 to 0.14.1.
- [Release notes](https://github.com/rustls/rcgen/releases)
- [Commits](https://github.com/rustls/rcgen/compare/v0.14.0...v0.14.1)

---
updated-dependencies:
- dependency-name: rcgen
  dependency-version: 0.14.1
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

flub committed:
Record stats and emit logs

flub committed:
still forgot a stat!

flub committed:
Start sending PATH_CIDS_BLOCKED (#106)

Sends and receives PATH_CIDS_BLOCKED and PATHS_BLOCKED correctly

dignifiedquire committed:
feat: add `upgrade` function for weak connection handle

flub committed:
quinn api: path access and status events

- Allows accessing the Path struct if you have a PathId

- Adds a Connection::path_events broadcast channel for the application
  to consume path events.

divagant-martian committed:
fix some path status bugs, improve path status story (#112)

* ensure path status is only queued one, and only when changed

* schedule sending the path status frame at a more appropriate instant

* fix clippy, save a line

* keep within the multipath world

* register the sent frame with retransmits

divagant-martian committed:
simplify getting the next path id (#113)

flub committed:
feat: add `upgrade` function for weak connection handle (#110)

flub committed:
try not to move this thing

flub committed:
quinn api: path access and status events (#111)

- Allows accessing the Path struct if you have a PathId

- Adds a Connection::path_events broadcast channel for the application
to consume path events.

dependabot[bot] committed:
build(deps): bump rcgen from 0.14.1 to 0.14.2

Bumps [rcgen](https://github.com/rustls/rcgen) from 0.14.1 to 0.14.2.
- [Release notes](https://github.com/rustls/rcgen/releases)
- [Commits](https://github.com/rustls/rcgen/compare/v0.14.1...v0.14.2)

---
updated-dependencies:
- dependency-name: rcgen
  dependency-version: 0.14.2
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

flub committed:
Abandon a path (#114)

Abandons a path, issues new MAX_PATH_ID and CIDs, closes connection if last path is abandoned.  Removes accepted reset tokens when a path is abandoned.

Ralith committed:
Configure qlog in TransportConfig

Ralith committed:
Record group ID with qlog metrics

Necessary for qlog streams shared between concurrent connections to be
intelligible.

Ralith committed:
Stabilize qlog feature

flub committed:
Tiny clean up of a comment that was in the wrong place

This comment being in the wrong place makes reading this code a bit
confusing.

flub committed:
Fix stateless reset comment (#115)

Ralith committed:
Factor qlog helpers out of Connection

Ralith committed:
Emit qlog PacketLost events

Ralith committed:
Rename CongestionMetrics to reflect the associated qlog structure

Ralith committed:
Correct pacing_rate metric units

draft-ietf-quic-qlog-quic-events-11 specifies bits per second.

Ralith committed:
Emit qlog PacketSent events

Ralith committed:
Emit qlog PacketReceived events

flub committed:
Do not track lost_packets twice

The number of lost packets is already tracked as path stats, there is
no need to duplicate it on the connection struct itself.  While access
was only ever in tests, the field would still be present in release
builds I think.

djc committed:
proto: fix NewConnectionId size bound

dependabot[bot] committed:
build(deps): bump serde_json from 1.0.140 to 1.0.141

Bumps [serde_json](https://github.com/serde-rs/json) from 1.0.140 to 1.0.141.
- [Release notes](https://github.com/serde-rs/json/releases)
- [Commits](https://github.com/serde-rs/json/compare/v1.0.140...v1.0.141)

---
updated-dependencies:
- dependency-name: serde_json
  dependency-version: 1.0.141
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump rcgen from 0.14.2 to 0.14.3

Bumps [rcgen](https://github.com/rustls/rcgen) from 0.14.2 to 0.14.3.
- [Release notes](https://github.com/rustls/rcgen/releases)
- [Commits](https://github.com/rustls/rcgen/compare/v0.14.2...v0.14.3)

---
updated-dependencies:
- dependency-name: rcgen
  dependency-version: 0.14.3
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump async-io from 2.4.1 to 2.5.0

Bumps [async-io](https://github.com/smol-rs/async-io) from 2.4.1 to 2.5.0.
- [Release notes](https://github.com/smol-rs/async-io/releases)
- [Changelog](https://github.com/smol-rs/async-io/blob/master/CHANGELOG.md)
- [Commits](https://github.com/smol-rs/async-io/compare/v2.4.1...v2.5.0)

---
updated-dependencies:
- dependency-name: async-io
  dependency-version: 2.5.0
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

djc committed:
proto: inline StepRng implementation from rand

djc committed:
Update to rand 0.9.2

dignifiedquire committed:
wip: allow server migrations

flub committed:
Allow opening path only if it does not yet exists (#117)

* Allow opening path only if it does not yet exists

* Hook this up to futures

All this is terrible.

* clippy

flub committed:
Improve ACK handling (#116)

* Immediately send ACKs for abandoned paths

This is required by § 3.4.3 from QUIC-MULTIPATH.

* Account for lost packets when dropping the path state

divagant-martian committed:
handle some edge cases in opening a path (#119)

* relly on existing paths for validation

* still send the challenge

* check the server gets the event too now that they validate the address as well

* remove path from frame spacn

This is already part of the spans that contain this one

* Set validation timer for new paths

* make sure validation failure removes the path

* check server events for failed validation test

* add client side failure test

* fix comment

* better docs

* add docs for the blackhole step

* set the previous log behaviour as default

Ralith committed:
Introduce unique path identifier

Ralith committed:
Use path ID to filter in-flight counter decrements

Identifying the path associated with a sent packet based on packet
number alone is only reliable if the path is still known, but we may
forget paths quickly e.g. when path validation fails or is abandoned.

We could alternatively retain information about abandoned paths so
long as they have in-flight packets, but this is simpler.

Ralith committed:
Remove dead arguments

flub committed:
Merge branch 'main' into multipath-quinn-0.11.x

flub committed:
Merge remote-tracking branch 'n0-computer/multipath-quinn-0.11.x' into multipath-quinn-0.11.x

flub committed:
merge quinn main (#120)

actually merge quinn main into an old commit of the multipath branch.
Then merge that old commit into the current head of the multipath
branch. Life is messy, and git doubly so.

divagant-martian committed:
fix open path timer issues (#121)

* set timer when setting the path challenge, still set the timer even if the path is validated

* modify comments about path challenge timeouts

* actually stop timer

Ralith committed:
debug_assert against inconsistent sent packet state

Ralith committed:
Check for in-flight data on-demand

Simpler and more robust than maintaining a separate running total.

Ralith committed:
Wake endpoint driver on rebind

Fixes connection loss and possible endpoint wedging.

Ralith committed:
Fix lints

Ralith committed:
Fix inconsistent in-flight handling of abandoned packets

dependabot[bot] committed:
build(deps): bump criterion from 0.6.0 to 0.7.0

Bumps [criterion](https://github.com/bheisler/criterion.rs) from 0.6.0 to 0.7.0.
- [Changelog](https://github.com/bheisler/criterion.rs/blob/master/CHANGELOG.md)
- [Commits](https://github.com/bheisler/criterion.rs/compare/0.6.0...0.7.0)

---
updated-dependencies:
- dependency-name: criterion
  dependency-version: 0.7.0
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump actions/checkout from 4 to 5

Bumps [actions/checkout](https://github.com/actions/checkout) from 4 to 5.
- [Release notes](https://github.com/actions/checkout/releases)
- [Changelog](https://github.com/actions/checkout/blob/main/CHANGELOG.md)
- [Commits](https://github.com/actions/checkout/compare/v4...v5)

---
updated-dependencies:
- dependency-name: actions/checkout
  dependency-version: '5'
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump windows-sys from 0.59.0 to 0.60.2

Bumps [windows-sys](https://github.com/microsoft/windows-rs) from 0.59.0 to 0.60.2.
- [Release notes](https://github.com/microsoft/windows-rs/releases)
- [Commits](https://github.com/microsoft/windows-rs/commits)

---
updated-dependencies:
- dependency-name: windows-sys
  dependency-version: 0.60.2
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump rustls-platform-verifier from 0.6.0 to 0.6.1

Bumps [rustls-platform-verifier](https://github.com/rustls/rustls-platform-verifier) from 0.6.0 to 0.6.1.
- [Release notes](https://github.com/rustls/rustls-platform-verifier/releases)
- [Changelog](https://github.com/rustls/rustls-platform-verifier/blob/main/CHANGELOG)
- [Commits](https://github.com/rustls/rustls-platform-verifier/compare/v/0.6.0...v/0.6.1)

---
updated-dependencies:
- dependency-name: rustls-platform-verifier
  dependency-version: 0.6.1
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump serde_json from 1.0.141 to 1.0.142

Bumps [serde_json](https://github.com/serde-rs/json) from 1.0.141 to 1.0.142.
- [Release notes](https://github.com/serde-rs/json/releases)
- [Commits](https://github.com/serde-rs/json/compare/v1.0.141...v1.0.142)

---
updated-dependencies:
- dependency-name: serde_json
  dependency-version: 1.0.142
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

gry8t committed:
Improved quinn::ConnectionStats docs

Co-authored-by: Benjamin Saunders <ben.e.saunders@gmail.com>

gretchenfrage committed:
docs(quinn): Revise SendStream::write docs

Aligns with 100 columns, attempts to improve clarity, takes inspiration
from tokio::io::AsyncWriteExt docs.

stormshield-damiend committed:
perf: factorize common options

stormshield-damiend committed:
perf: used buffered writer for qlog stream

stormshield-damiend committed:
perf: factorize init of tracing

stormshield-damiend committed:
perf: factorize socket binding

abonander committed:
chore: add test cases for `set_send_window`

djc committed:
quinn: drop async-std support

stormshield-damiend committed:
cov: fix qlog feature name

stormshield-damiend committed:
perf: support SI units for the various size parameters

stormshield-damiend committed:
perf: allow setting send / receive connection or stream windows

stormshield-damiend committed:
bench: refactor parse_byte_size

stormshield-damiend committed:
bench: fix clap help

Dreamacro committed:
perf: add max udp payload size option

dependabot[bot] committed:
build(deps): bump url from 2.5.4 to 2.5.7

Bumps [url](https://github.com/servo/rust-url) from 2.5.4 to 2.5.7.
- [Release notes](https://github.com/servo/rust-url/releases)
- [Commits](https://github.com/servo/rust-url/commits)

---
updated-dependencies:
- dependency-name: url
  dependency-version: 2.5.7
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump actions/setup-java from 4 to 5

Bumps [actions/setup-java](https://github.com/actions/setup-java) from 4 to 5.
- [Release notes](https://github.com/actions/setup-java/releases)
- [Commits](https://github.com/actions/setup-java/compare/v4...v5)

---
updated-dependencies:
- dependency-name: actions/setup-java
  dependency-version: '5'
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
build(deps): bump serde_json from 1.0.142 to 1.0.143

Bumps [serde_json](https://github.com/serde-rs/json) from 1.0.142 to 1.0.143.
- [Release notes](https://github.com/serde-rs/json/releases)
- [Commits](https://github.com/serde-rs/json/compare/v1.0.142...v1.0.143)

---
updated-dependencies:
- dependency-name: serde_json
  dependency-version: 1.0.143
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

Dreamacro committed:
perf: move client and server on one binary

stormshield-damiend committed:
perf: rename perf binary to quinn-perf to prevent name clash with linux perf

djc committed:
Ignore async-std advisory for now

djc committed:
Upgrade to fastbloom 0.14

djc committed:
Bump quinn to 0.11.9

djc committed:
Update semver-compatible dependencies

djc committed:
Bump quinn-udp to 0.5.14

flub committed:
Only allow server to migrate before its address is validated

Once we have validated the server's address during the handshake we
should no longer allow the serers's address to migrate.  Before that
we're just accepting any migrations.

This also adds the logic to consider the server's address validated in
a retry packet if possible.

flub committed:
clippy

flub committed:
update deny.toml with whatever current silly shenanigans

flub committed:
use same syntax as above

flub committed:
If only I could write if conditions

flub committed:
semicolon

flub committed:
allow server migrations (#118)

To support sending initial packets to many destinations and accepting
the first one received back, we need to allow the server address to
migrate.

Once we have validated the server's address during the handshake we
no longer allow the server's address to migrate.

This also adds the logic to consider the server's address validated in
a retry packet if possible.

djc committed:
Remove more traces of async-std

divagant-martian committed:
add public api for observed address with multipath (#122)

* add public api for observed address with multipath

* test

* add temp todo

* fix weird phrasing in docs, add todo

* fix ordering of events, complete test

* fix lint failure

* partially fix runtime issue

* fix lint

* add api test

* fix timing issue

* remove unnecesary variable

* cleanup

divagant-martian committed:
Fix open path api (#123)

* fix open path api

* simplify code

divagant-martian committed:
prevent servers from opening paths, on the client side (#124)

* prevent servers from opening paths, on the client side

* reduce noice, fix comment

flub committed:
Make PathError and actual Error

flub committed:
Make PathError and actual Error (#125)

divagant-martian committed:
Merge commit '5a45f0846c32f2ca76a078165bf61211033ac26e' into multipath-quinn-0.11.x

divagant-martian committed:
Merge commit '54b39d1c004a56ca71d8d18f12f6f232f13a7e5a' into multipath-quinn-0.11.x

# Conflicts:
#	quinn-proto/src/connection/mod.rs
#	quinn-proto/src/connection/spaces.rs

divagant-martian committed:
Merge commit 'd579a38fab6b8a00f3925e966b02787486047676' into multipath-quinn-0.11.x-b2b930a-merges

divagant-martian committed:
Merge commit 'c5b556736cf8e5297b3d52eb58144318ca490dd6' into multipath-quinn-0.11.x-b2b930a-merges

# Conflicts:
#	.github/workflows/rust.yml
#	quinn-proto/src/connection/spaces.rs

divagant-martian committed:
Merge commit '947d45a26e05f9ece6247c0a860aa9a98262759d' into multipath-quinn-0.11.x-b2b930a-merges

divagant-martian committed:
Merge commit 'ca74bd22b766235a03362c2291fc6401e728aaf9' into multipath-quinn-0.11.x-b2b930a-merges

divagant-martian committed:
Merge commit '14d7b69892eec5e83cc0290869008102f454f6aa' into multipath-quinn-0.11.x-b2b930a-merges

# Conflicts:
#	quinn-proto/src/connection/mod.rs
#	quinn-proto/src/connection/paths.rs

divagant-martian committed:
Merge commit '4625fe084ca37b5c9f057e6e1113b5f94bb5e8e3' into multipath-quinn-0.11.x-b2b930a-merges

# Conflicts:
#	quinn-proto/src/connection/mod.rs

divagant-martian committed:
Merge commit 'c1b181d17bd71a2846078893b9f91892fbfa8cbb' into multipath-quinn-0.11.x-b2b930a-merges

divagant-martian committed:
Merge commit 'bb46d09f00c14ddedca9c88dc961225ac7dd4a4e' into multipath-quinn-0.11.x-b2b930a-merges

# Conflicts:
#	.github/workflows/book.yml
#	.github/workflows/rust.yml

divagant-martian committed:
Merge commit '3194714f6fa1c78108b2a6353bd3e2af95ce1152' into multipath-quinn-0.11.x-b2b930a-merges

divagant-martian committed:
Merge commit 'b2b930a0662b18b2e351264a21e175478bb3c3f1' into multipath-quinn-0.11.x-b2b930a-merges

divagant-martian committed:
fix docs from removed field

divagant-martian committed:
feat: Update multipath to latest released quinn (b2b930a) (#128)

divagant-martian committed:
Revert Dig's merge fixed to bring back const

This reverts commit 44a3dc6e454b1dab087fa5ce74d072c9f3b0fdf1.

divagant-martian committed:
fixes to make everything work together

divagant-martian committed:
cherry-pick Phillip's changes from quinn/main (#129)

This also includes removing async-std support. These are commits on top
of which Phillip's ones are made

flub committed:
Add OpenPath::path_id and some docs (#127)

* Add OpenPath::path_id and some docs

* Try and word this a little better

flub committed:
Allow server migration before authenticated handshake packets (#130)

The previous logic does not work: the client has the path validated
immediately, even before any packet is sent.

Now we need to have received an authenticated handshake packet before
we stop allowing migration.  This exposes us to a little more off-path
attacking, but in terms of processing all the earlier packets: only
the correct server can generate an authenticated handshake packet IIUC.

flub committed:
Allow the WeakConnectionHandle to be cloned (#131)

flub committed:
Expose the remote address of a path (#133)

flub committed:
Fix WeakConnection::upgrade (#132)

* Fix WeakConnection::upgrade

It created a ConnectionRef that did not increment the manually-tracked
reference count inside the Arc<ConnectionInner>.  This triggered the
ConnectionRef's Drop impl to actually drop the connection.

This works around this by using the fact that the ConnectionRef's
Clone impl does know how to manipulate the reference count in a way
that works together with the ConnectionRef's Drop.  I think this is
preferred over manipulating the ref_count directly in the
WeakConnectionHandle since that code is closer by.

* Nicer implementation

flub committed:
upgrade rustls to 0.23.29 to match iroh (#134)

flub committed:
Trace log the OBSERVED_ADDR frames (#135)

This in the same style as the other frames are logged.  It's very
confusing to not have a trace log of this.

flub committed:
Make Path impl fmt::Debug (#138)

flub committed:
Only send an MTUD if there is an active CID, fixes a panic (#139)

flub committed:
Allow setting the status of a path (#140)

flub committed:
fix span field name (#142)

If only tracing would be type-safe.  But yeah, strings are fiiiine

flub committed:
Bump the max padding allowe in GSO batches (#143)

When building packets in GSO batches the batch is terminated if too
many padding bytes would be needed to continue this batch.  This was
chosen fairly arbitrarily to be 16 bytes.

However when you have multipath enabled and allow a half-decent number
of simulatneous paths to be opened, you have to issue 5 *
initial_max_path_id PATH_NEW_CONNECTION_ID frames.  This quickly
spills over into multiple datagrams, especially since this happens
early in the connection when the PMTU has not yet been discovered.

CIDs however, are 32 bytes by default.  So if such a frame does no
longer fit in a datagram you would easily need more than 16 bytes of
padding.  And in that case the GSO batch would break.  Even though all
you're doing is sending a series of datagrams full of CIDs.

So tweaking this number to 32 allows us to build these packes without
breaking the GSO batch.

flub committed:
Fix remote addresses in paths (#141)

* Fix remote addresses in paths

- Remove quinn_proto::Connection::remote_address.  You should use
  path_remote_address now.

- Change quinn::Connection::remote_address to return *any* remote
  address from any of the open paths.

- Fix opening paths on IPv4 addresses, when the endpoint supports
  IPv6.

  When the endpoint supports IPv6 it uses IPv4-mapped IPv6 addresses
  for IPv4 remotes.  This is enforced when establishing a connection,
  and quinn-udp automatically does this for incoming datagrams.
  However when opening a path we should do the same mapping.
  Otherwise we will end up with an IPv4 remote address in the PathData
  and incoming datagrams will come from an IPv6 address and be
  dropped.

* fix tests

divagant-martian committed:
cleanup frame.rs ToDos and unwraps (#144)

* remove obsolete dead code annotation

* add missing docs for coding and encoding path related frames

* make `RetireConnectionId` max size into consts

* make `NewConnectionId` max size into consts

* remove unwraps comming from n0 members in frame.rs

divagant-martian committed:
add missing path stats in debug impl, remove obsolte comment (#145)

flub committed:
remove redundant logging (#147)

I messed this up earlier, this is already logged.

divagant-martian committed:
cleanup use of PathId constants and access to inner (#148)

* use PathId::ZERO whenever possible

* avoid accesing PathId's inner in general

* fmt

* prefer PathId::MAX over max construction

divagant-martian committed:
remove TODOs since the current state seems acceptable (#149)

divagant-martian committed:
enforce MAX_PATH_ID and abandoned paths check for new CIDs (#150)

flub committed:
Switch to released rustls (#152)

divagant-martian committed:
Fix connection's mtu calculation (#151)

* calculate this connection's mtu as the minimum across all paths

* add TODO to check buffer reserving

* improve docs

* apply clippy fixes

* more docs tweaks

* second perspective on TODO

* spelling

* address TODO: reserving is already done in poll_transmit with the appropiate sizes

divagant-martian committed:
add AddAddress frame (#154)

divagant-martian committed:
add PunchMeNow frame (#155)

divagant-martian committed:
add RemoveAddress frame (#156)

flub committed:
Fix check for max permitted PathId in PATH_CIDS_BLOCKED frame (#157)

Also sneak in a unrelated logging improvement.

divagant-martian committed:
add transport parameter (#158)

* add nat_traversal field to TransportParameters

* add nat traversal's concurrency limit field to TransportConfig

* add nat_traversal TransportParameterId

* make field private to prevent zero values, reject 0 always

* enforce 0-rrt data matches

divagant-martian committed:
nat traversal requires multipath (#159)

* nat traversal requires multipath

* allow crate access, not ideal, but unavoidable

* satisfty MSRV limits

* pub all the way to the top

* update wasm-bindgen-test

flub committed:
Improve docs (#161)
